### ISSUE TITLE
Main window flips instead of being shown with click on notification icon

### ISSUE THREAD

**Issue by irgendsontyp** [keepassxreboot/keepassxc#2088]
## Expected Behavior
The main window shall be shown and stay visible.

## Current Behavior
The main window is shown for a very short time (less than a second) and suddenly disappears.

## Steps to Reproduce (for bugs)

1. Ensure that in Tools -> Settings -> General -> General the options "Show a system tray icon" and "Hide window to the system tray when minimized" are checked.
2. Close KeePassXC.
3. Start KeePassXC.
4. Minimize KeePassXC via the panel (not the notification icon!)
5. Click the notification icon.

## Debug Info

KeePassXC - Version 2.3.3
Revision: 0a155d8

Libraries:
- Qt 5.11.1
- libgcrypt 1.8.3

Operating system: Arch Linux
CPU architecture: x86_64
Kernel: linux 4.17.3-1-ARCH

Enabled extensions:
- Auto-Type
- Browser Integration
- Legacy Browser Integration (KeePassHTTP)
- SSH Agent
- YubiKey

**Comment by droidmonkey** [comment]
What window manager are you using?

**Comment by InTheMorning** [comment]
My details are exactly the same, save for the extensions where I am only using Auto-Type
and Browser Integration.

I am using i3wm.  Killing the window using the wm and bringing it back returns an empty window.  Need to restart keepass to get the window back.  

Toggling using the systray icon works as expected.

**Comment by irgendsontyp** [comment]
@droidmonkey I am using xfwm4, version 4.12.4. My desktop environment is xfce4. Everything is up to date with the arch repositories.

**Comment by droidmonkey** [comment]
Both of your descriptions sound like the problem lies with Qt. We have had reports of negative interactions with i3 and xfce (although relegated to the snap version).

**Comment by irgendsontyp** [comment]
This is interesting. It seems to happen only when the KeePassXC window is maximized before minimizing it. 

Also I found out something strange: Adding the line

`std::cout << "isMinimized(): " << static_cast<QWindowStateChangeEvent*>(event)->oldState() << ", isMaximized(): " << windowState() << std::endl;`

in src/gui/MainWindow.cpp after line 829 and following the instructions in my initial comment in this issue, I get that output:

`isMinimized(): 1, isMaximized(): 1`.

So the window is minimized and maximized at the same time, if you believe Qt. There indeed seems to be something wrong with Qt (or the window manager? I'm not deep enough into that kind of stuff). Maybe playing around with the internal states of the Qt classes leads to something.

**Comment by irgendsontyp** [comment]
I guess it's not as strange as I thought that isMinimized() returns 1 and isMaximized() does the same.

But I found another interesting thing: If you [modify MainWindow.cpp](https://github.com/irgendsontyp/keepassxc/commit/f05850456fe7111f63413fe3a994004342304c22) to use a timer for maximizing the main window, the flip goes away in **certain cases**. The lower the timeout is, the bigger the chance gets for the flip to appear. Try it yourself: Having the timeout at 200, the flip does not appear. Having it at 1, it flips.

Maybe there is still some work going on in the background in the Qt libs which finishes after 200 ms but not after 1 ms. Currently I'm building Qt 5 from source to see if I can debug the issue there.

**Comment by droidmonkey** [comment]
Yes this could very well be a race condition or a "double trigger" scenario where the delay prevents the second trigger.

**Comment by joanbm** [comment]
It indeed looks like it has the same underlying case as bug #1595 that was already closed, though it's strange that this case wasn't detected during the fix, since I remember doing quite a bit of testing of cases like those, and it passed the testing by others during the PR as well. Perhaps there has been some change in Qt to expose this.

I can reproduce the bug on the latest testing Arch with XFCE, but like in the original bug, some times it consistently reproduces, other times it doesn't ever reproduce, not sure why.

The comment on `hideWindow()` sheds some light on this. Based on this, I tried the following workaround on `changeEvent(...)`:

```
diff --git a/src/gui/MainWindow.cpp b/src/gui/MainWindow.cpp
index 5ef48f00..2d55f121 100644
--- a/src/gui/MainWindow.cpp
+++ b/src/gui/MainWindow.cpp
@@ -715,6 +715,9 @@ void MainWindow::changeEvent(QEvent* event)
         if (isTrayIconEnabled() && m_trayIcon && m_trayIcon->isVisible()
             && config()->get("GUI/MinimizeToTray").toBool()) {
             event->ignore();
+#if defined(Q_OS_LINUX) || defined(Q_OS_MAC)
+            setWindowState(windowState() & ~Qt::WindowMinimized);
+#endif
             QTimer::singleShot(0, this, SLOT(hide()));
         }
 ```

This seems to fix the bug when restoring the window on my system, except that it looks like it causes a different race condition while handling the event, and around 20% of the time, the window doesn't even minimize and starts glitching.

**Comment by irgendsontyp** [comment]
As I think the best way is not to find workarounds but rather fix the root cause, I created a bug in the Qt bug tracker: https://bugreports.qt.io/browse/QTBUG-69375.

**Comment by irgendsontyp** [comment]
I updated the bug to clarify what exactly is the problem. My initial description was a bit misleading because it did not emphasize what the real problem is.

The strange thing is that in the attached QtCreator example project, increasing the singleShot timer value in mainwindow.cpp, line 18, still causes the window to flip (which is different from KeePassXC). Did I miss something in the code? I'd appreciate you guys having a look into the bug too ensure both my description and code are valid for the bug.

**Comment by droidmonkey** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [#2088 - missing flags on maximize from tray](https://trello.com/c/TihmKKBI/7-2088-missing-flags-on-maximize-from-tray)

**Comment by irgendsontyp** [comment]
> ![](https://camo.githubusercontent.com/193766a3b9959c5f4ed5cd8cf3251d015e839d51/68747470733a2f2f6769746875622e7472656c6c6f2e73657276696365732f696d616765732f6d696e692d7472656c6c6f2d69636f6e2e706e67) [#2088 - missing flags on maximize from tray](https://trello.com/c/TihmKKBI/7-2088-missing-flags-on-maximize-from-tray)

I get some kind of permission denied error. What's supposed to be behind this link?

**Comment by droidmonkey** [comment]
This is fixed in 2.5.3
